### PR TITLE
Add Contents and Config

### PR DESCRIPTION
**Pull request by blink1073**
Adds the Contents and Config services, as well as a swagger REST API  [definition](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/blink1073/phosphor-notebook-1/other-services/rest_api.yaml#!/contents).

Follows https://github.com/ipython/ipython/wiki/IPEP-27:-Contents-Service and https://github.com/jupyter/notebook/blob/master/notebook/services/contents/handlers.py.

### CODE REVIEW

**Comment by Carreau** [comment]
The contents manager should be swappable, can we have an interface for it ? I have that lying around :+1: 

```
/**
 * Interface that a content manager should implement
 **/

import notebook_model = require("notebook_model")
import Notebook = notebook_model.Notebook;

export interface IContents {
    get(path:Path, type:FileType, options:Object):any
    new_untitled(path:Path, options:Object):Notebook
    delete(path:Path):void
    rename(path: Path, new_path: Path):Promise<any>
    save(path: Path, model: any, options?:any):Promise<any>
    list_contents(path: Path, options: any):Promise<any>
    copy(path: Path, model: any):Promise<any>
    create_checkpoint(path: Path, options: any):any
    restore_checkpoint(path: Path, checkpoint_id: CheckpointId, options: any):Promise<any>
    list_checkpoints(path: Path, options: any):any
}
```

And I also have these aliases: as in some places (mixed content manager) there is metaprogamming and having "String" everywhere make it hard to follow. 

```
export interface Path extends String {}
```

**Comment by sccolbert** [comment]
+1 on making this an interface. As soon as we start getting into things we want to be swappable or pluggable, we need to have well-defined interfaces which are as simple as possible. We should also create an interface token for them so that they can be used with Phosphor's dependency injection and bootstrapping system:
https://github.com/phosphorjs/phosphor/tree/master/src/di
https://github.com/phosphorjs/phosphor/blob/master/src/shell/ishellview.ts#L89
https://github.com/phosphorjs/phosphor/blob/master/src/shell/bootstrapper.ts#L128
https://github.com/phosphorjs/phosphor/blob/master/examples/chatapp/chat/roomsplugin.ts#L108

-1 on the empty extension of String. That gains (functionally) nothing in typescript. You could maybe make the case for extra documentation, but `path: string` is sufficient and self-explanatory IMHO.

**Comment by Carreau** [comment]
When you get something like:

```
private _route_function(method_name:string, arg_types:ArgType[], return_type, args) {
        return this._filesystem.then((filesystem) => {
            if (arg_types.length == 0 || arg_types[0] != ArgType.PATH) {
                // This should never happen since arg_types is hard coded below.
                throw 'unexpected value of arg_types';
            }
            var root = this.get_fs_root(filesystem, args[0]);

            if (root === '') {
                if (method_name === 'list_contents') {
                  return {'content': this._virtual_fs_roots(filesystem)};
                } else {
                  throw 'true root directory only contains mount points.';
                }
            }

            for (var i = 0; i < args.length; i++) {

                args[i] = this.from_virtual(root, arg_types[i], args[i], this._config.data['mixed_contents']['schema']);
            }
            var contents = filesystem[<string>root];
            return contents[method_name].apply(contents, args).then(
                $.proxy(this._to_virtual, this, root, return_type));
        });
    }
```

The type inference that the methods like `from_virtual` return Path or string, is useful to catch error in the number/type of arguments put in args.

But sure we can start with String, but when you start manipulating abstract functions signature, more typing might be worth it.

**Comment by sccolbert** [comment]
I'm not really following you. Can you give a simpler example of what you are talking about? I don't understand why you are trying to distinguish between `Path`, `String`, and `string`.

**Comment by Carreau** [comment]
basically I have 2 contents managers, The one in this PR, and the one that [talk to google drive](https://github.com/jupyter/jupyter-drive/blob/master/jupyterdrive/ts/drive-contents.ts) I want to virtually have an hybrid content manager that expose the one in this PR under `/local/` and the google one under `/gdrive/` (which I have [here](https://github.com/jupyter/jupyter-drive/blob/master/jupyterdrive/ts/mixed-contents.ts)) 

So I'm using meta-programming (decorators soon I guess) to strip down/prepend `/local/`,`/gdrive/` when the args/returns values are paths, and I would prefer the typing system to warn me when I do that on a non-path.

So typically I want things like [that](https://github.com/jupyter/jupyter-drive/blob/master/jupyterdrive/ts/mixed-contents.ts#L247-L250), to complain if I want to apply a decorator to a function like `foo(a:Path, b:String)`, but not on `foo(a:Path, b:Path)`.

**Comment by sccolbert** [comment]
since `Path` is an empty extenstion of `String`, there is not difference between them as far as TypeScript is concerned. So, if I understand you correctly, you won't be able to the compiler error you want.

**Comment by sccolbert** [comment]
Does [this example](http://www.typescriptlang.org/Playground#src=%0Ainterface%20Path%20extends%20String%20%7B%20%7D%0A%0A%0Afunction%20doThingToPathFn%28fn%3A%20%28path%3A%20Path%29%20%3D%3E%20void%29%3A%20void%20%7B%0A%09%0A%7D%0A%0A%0Afunction%20foo1%28path%3A%20string%29%3A%20void%20%7B%20%7D%0A%0Afunction%20foo2%28path%3A%20String%29%3A%20void%20%7B%20%7D%0A%0Afunction%20foo3%28path%3A%20Path%29%3A%20void%20%7B%20%7D%0A%0A%0AdoThingToPathFn%28foo1%29%3B%0AdoThingToPathFn%28foo2%29%3B%0AdoThingToPathFn%28foo3%29%3B) reflect what you are wanting to do?

This doesn't throw an error since all the types are structurally compatible.

**Comment by blink1073** [comment]
@jasongrout, can you help out with the REST API, and I'll fill in the functions here appropriately?

**Comment by jasongrout** [comment]
Sure, on Monday.

**Comment by Carreau** [comment]
@sccolbert Roughly, [update version](http://www.typescriptlang.org/Playground#src=%0Ainterface%20Path%20extends%20String%20%7B%20%7D%0A%0A%0Afunction%20doThingToPathFn%28fn%3A%20%28path%3A%20Path%29%20%3D%3E%20void%29%3A%20void%20%7B%0A%09%0A%7D%0A%0A%0Afunction%20foo1%28path%3A%20string%29%3A%20void%20%7B%20%7D%0A%0Afunction%20foo2%28path%3A%20String%29%3A%20void%20%7B%20%7D%0A%0Afunction%20foo3%28path%3A%20Path%29%3A%20void%20%7B%20%7D%0A%0A%0AdoThingToPathFn%28foo1%29%3B%0AdoThingToPathFn%28foo2%29%3B%0AdoThingToPathFn%28foo3%29%3B) , though I don't get why it does not complain. Locally if I replace some `Path` annotation with `string` I get some `jupyterdrive/ts/mixed-contents.ts(158,36): error TS2345: Argument of type 'Path' is not assignable to parameter of type 'string'.` which made me probably believe that the type system is stricter. I might be wrong, I'll try to come up with the minimal example. But lets move forward with `string` if it suit you I don't want to slow things down, and you have more experience with TS than I.

**Comment by sccolbert** [comment]
@Carreau I don't see any difference between your updated example and my original. But, yeah type compatibility can be tricky at first. It helps to not think about TypeScript types as you would C/C++ types. TypeScript is much more like duck typing: if something has the same set of properties and methods as the target type, it's considered compatible. Maybe [this example](http://www.typescriptlang.org/Playground#src=%0Aclass%20Foo%20%7B%0A%09a%3A%20number%3B%0A%7D%0A%0A%0Aclass%20Bar%20%7B%0A%09a%3A%20number%3B%0A%09thing%28%29%3A%20void%20%7B%20%7D%0A%7D%0A%0A%0Aclass%20Baz%20%7B%0A%09a%3A%20number%3B%0A%09other%28%29%3A%20void%20%7B%20%7D%0A%7D%0A%0A%0Avar%20foo%20%3D%20new%20Foo%28%29%3B%0Avar%20bar%20%3D%20new%20Bar%28%29%3B%0Avar%20baz%20%3D%20new%20Baz%28%29%3B%0A%0Avar%20foo2%3A%20Foo%3B%0Avar%20bar2%3A%20Bar%3B%0Avar%20baz2%3A%20Baz%3B%0A%0A%2F%2F%20everything%20is%20compatible%20with%20Foo%20even%0A%2F%2F%20though%20they%20don't%20share%20a%20common%20class%0Afoo2%20%3D%20bar%3B%0Afoo2%20%3D%20baz%3B%0A%0A%2F%2F%20not%20compatible%0Abar2%20%3D%20foo%3B%0Abar2%20%3D%20baz%3B%0A%0A%2F%2F%20not%20compatible%0Abaz2%20%3D%20foo%3B%0Abaz2%20%3D%20bar%3B%0A) will help demonstrate.

**Comment by blink1073** [comment]
I finished another pass-through to rectify the Swagger API, the TypeScript class, the IPEP, and the Tornado class.  

Two open questions: 
- Should there be an `upload` method in Typescript?  Is that part of the API used anywhere?  Right now we are only using PUT for `save`.  https://github.com/ipython/ipython/wiki/IPEP-27:-Contents-Service#upload-a-file
- Do we need to use the header Location in Typescript?  All of the models contain a path in them.  This would affect `rename`, `copy`, `newUntitled`,`save`, and `createCheckPoint`.

**Comment by blink1073** [comment]
Ah, `createCheckPoint` does not have a path in the model, but the rest do.

**Comment by blink1073** [comment]
Moved to: https://github.com/jupyter/jupyter-js-services/pull/17.
